### ISSUE TITLE
FR: petty url slug for published note

### ISSUE THREAD

**Issue by andrey-jef** [oleeskild/obsidian-digital-garden#4]
This is a feature request. 

Looking at your demo url, I wonder if you can make an option in the plugin to replace 'space' by hyphen '-' instead of '%20', and all lowercase. 

Or even better if I can assign a key in frontmatter to define a slug for specific notes, while the rest will be generated automatically by default based on filename. 

https://ole.dev/notes/Set%20up%20your%20own%20digital%20garden

**Comment by oleeskild** [comment]
Thanks for showing an interest in the project!
		
Can I ask you the use case you envisioned for this? The reason I ask is because it would introduce extra complexity adding this feature, simply because it would break the obsidian links ([[Some Other Note]]) as it would no longer be possible to just redirect user to the url with that title. This means that you would need to check, or keep a mental model in your head whether you can link to a note with normal Obsidian linking or not. 
		
Perhaps a solution would be to somehow add an 'alias', where you could reach the note with both the slug path, and the full title path.

Would your problem be solved by me adding a new command which lets you copy the link to the current note in obsidian to your clipboard? Or is it that you simply don't like the aesthetic of using encoded spaces in the URL? Or something completely else?

**Comment by andrey-jef** [comment]
Thanks for taking time to explain how the FR could break then internal links between published notes within an Obsidian vault.

- My reason for replacing '%20' as hyphen is simply aesthetic as you pointed out. I admit that this is the least important factor for a personal garden, but [this article from ahrefs](https://ahrefs.com/blog/seo-friendly-urls/) about seo-friendly url might resonate with what we are discussing
- Secondly for the 'lower case' part, since you deploy on Netlify. I had experience of an error where [Netlify makes all urls lower case](https://www.jvt.me/posts/2019/11/11/gotcha-netlify-lowercase/), when I used Netlify to publish my notes with [Dendron](https://www.dendron.so/), in which urls are case-sensitive. So I doubt whether the case-sensitive of url from your plugin will pose any conflict?

Your proposed solution about the 'slug-alias' in frontmatter where a note can be reached by both the 'slug' or its 'full title path' sounds promising.

Currently, I believe that your plugin is the most simple and intuitive-to-use alternatives of `Obsidian Publish` in gratis, to publish notes from an Obsidian's vault

**Comment by oleeskild** [comment]
After giving it some thought, I've concluded that slugifying all the URLS is the way to go. I've updated the digitalgarden template so that all links will be a slugified title using
this package https://www.npmjs.com/package/@sindresorhus/slugify. This, among other things, will convert spaces to dashes and uppercase to lowercase. 

I also liked the idea of supporting permalinks, and came up with a solution that doesn't break the linking between notes. 

Both of these changes are done in the [template](https://github.com/oleeskild/digitalgarden) . You don't need to update the plugin. Only the template deployed to Netlify. 

It works like this:
If you add
```
---
permalink: "mynote/"
---
```
to the frontmatter of your file, this will be the URL of your note. Additionally all regular obsidian links to that file will also use this link. 
If you ever decide to delete the permalink and use the default behavior, all the links into this note will automatically be updated as soon as you publish the changed note. 
NOTE: It is important to always use the trailing "/". Without it, most browsers will download the file instead of rendering it. 
You can also use multiple levels deep paths like this
```
---
permalink: "category/2022/mynote/"
---
```
It would be great if you could test it and report back if you had any issues!

In order to use it you will need to update your template used in Netflify. 
I think the easiest way is to simply use the deploy button once again from [this repo](https://github.com/oleeskild/digitalgarden) and publish a new site to Netlify. Then re-publish all your notes to the new repo, before deleting the old one.

I'm working on a way to make the update process of a template available at the press of a button from the plugin, but for now you'll unfortunately have to do it manually.

**Comment by andrey-jef** [comment]
Great update. I've successfully applied the updated slug and permalink features. Thanks. 

Please tell me if i need to open a new issue. Since the following discussion diverges from the issue's title

With experience of the plugins in Obsidian like _admonition_ and _breadcrumbs_, they had to change their use of keywords in frontmatter to avoid possible future conflicts 
- breadcrumbs changed from `parent` to `bc-parent`
- admonition added `ad-` to all of its key

I wonder whether it would be better if you use a key which is specific uniquely to your plugin 
- 'permalink' could possibly be 'odg-permalink'
- about the `tags: gardenEntry`, I'd love to learn why you use this key-value pair instead of a dedicated key like `odg-publish: true`

**Comment by andrey-jef** [comment]
> * about the `tags: gardenEntry`, I'd love to learn why you use this key-value pair instead of a dedicated key like `odg-publish: true`

Reading more about [11ty.js](https://11ty.rocks/posts/going-beyond-static-with-eleventy/), I understood that you have to use `tags: gardenEntry` to access the `gardenEntry` collection within the template.

**Comment by oleeskild** [comment]
Great input! Thats a good point. The reason for using permalink is the same as tags: It's what 11ty uses. 
I've published a [new version](https://github.com/oleeskild/obsidian-digital-garden/releases/tag/1.4.0) where I've created plugin-specific keywords, which will be converted to tags and permalinks attributes before it's processed by 11ty.
 
 You can now use dg-permalink: "mynote" and dg-home: true. Also the permalink no longer needs to end with a "/".

**Comment by andrey-jef** [comment]
Aha. Quite a joy to read your updated readme and main.ts changes

Reading the main.ts, i think it still has an edge case where user accidentally use your plugin and has a note with the `tags: gardenEntry`, but without the intention to publish that note. Though we would not see the conflict soon. 

Secondly, I would like to clarify whether I understand the publishing workflow with the plugin. Please correct me if I were wrong. 
- at initial, add `dg-home: true` in the frontmatter of the note that I want to be my homepage
- Then run the command `Digital Garden: Publish Note` for every note that I want to publish

If it was the case, on the end-user viewpoint, I'd love to have a `dg-publish: true` in every note that I want to publish. And I can run the command `Digital Garden: Publish Note` only once, to push the changes and publish all notes with `dg-publish: true`. The benefits for end user with this approach are as follows. 
- user has a clear indicator in each markdown file, to know which note was published
- user doesn't need to run the command `Publish Note` multiple times when he wants to update the contents of a note or publish multiple notes at the same time

I'm eager to learn what you would think, and how complex is it to implement.

**Comment by oleeskild** [comment]
You understood the workflow correctly.

I was thinking about how I could somehow keep an overview of which notes are published or not. Using an attribute in frontmatter, as you suggest, is a great solution. Then you could combine that with dataview and do a query like "list where dg-publish = true" to get an overview over every note that is published. 

I think I'll keep the "Publish Note" command, so that it is still possible to only publish one note at a time. But it will require the user to have specified the "dg-publish: true" attribute. If not there I'll show a notice telling the user that it is missing. Hopefully this will make those who already uses it aware of the breaking change. 

Additionally I will also introduce a new command called "Publish All Notes" which will publish all notes containing the attribute dg-publish: true. 

Making the dg-publish: true mandatory is a breaking change. But I think that's fine considering that the plugin hasn't yet entered the official plugin list yet. I've implemented the feature, but I need to do some more testing and update the docs before I publish it.

Thanks for some really good suggestions on how to make the plugin better! 😄

**Comment by andrey-jef** [comment]
Thanks for taking time to discuss and implement. 

About the name of command, I have suggestions. 
- `Publish multiple notes` instead of `Publish all notes`
  - because `Publish all notes` might be confused that all the notes will be published
- `Publish single note` vs `Publish note`
  - describe the action more strictly

**Comment by oleeskild** [comment]
I agree! New release with these features implemented has been published.
